File: couchapps/editions/covers/views/by_id/map.js
function(doc) {
    var last_modified;
    
    function extend(d1, d2) {
        for (var k in d2) {
            d1[k] = d2[k];
        }
    }
    
    function extend_list(list1, list2) {
        for (var i=0; i<list2.length; i++) {
            list1.push(list2[i]);
        }
        return list1;
    }
    
    function get(d, key, default_value) {
        return d[key] || default_value;
    }
    
    if (!doc.covers || doc.covers.length == 0) {
        return;
    }
    
    if (doc.last_modified && doc.last_modified.value) {
        last_modified = doc.last_modified.value;
    }
        
    var cover_value = {"key": doc.key, "cover": doc.covers[0], 'last_modified': last_modified};
    
    var d = {};
    
    doc.identifiers && extend(d, doc.identifiers);

    d.lccn = get(doc, "lccn", []);
    d.oclc = get(doc, "oclc_numbers", []);
        
    d.isbn = [];
    doc.isbn_10 && extend_list(d.isbn, doc.isbn_10);
    doc.isbn_13 && extend_list(d.isbn, doc.isbn_13);
    
    for (var i=0; i<d.isbn.length; i++) {
        d.isbn[i] = d.isbn[i].replace(/[ -]+/g, "");
    }
    
    for (var name in d) {
        var value = d[name];
        for (var i=0; i<value.length; i++) {
            emit([name, value[i]], cover_value);
        }
    }
    
    doc.ocaid && emit(["ocaid", doc.ocaid], cover_value);

    var olid = doc['key'].split("/")[2];
    emit(["olid", olid], cover_value);
}

File: couchapps/editions/ids/views/by_id/map.py
def map(doc):
    last_modified = doc.get("last_modified", {}).get("value", "")
    covers = doc.get("covers") or [-1]
    
    cover_value = {"cover": covers[0], 'last_modified': last_modified}
    
    d = {}
    d.update(doc.get("identifiers", {}))
    
    d.update({
        "isbn": doc.get("isbn_10", [])  + doc.get("isbn_13", []),
        "lccn": doc.get("lccn", []),
        "oclc": doc.get("oclc_numbers", []),
    })
    
    d['isbn'] = [isbn.upper().replace("-", "") for isbn in d['isbn']]
    
    d['ia'] = [s[len("ia:"):] for s in doc.get("source_records", []) if s.startswith("ia:")]
    if 'ocaid' in doc and doc['ocaid'] not in d['ia']:
        d['ia'].append(doc['ocaid'])
    
    d['olid'] = [doc['key'].split("/")[-1]]
    
    f = open("/tmp/couch.log", 'a')
    print >> f, doc['key'], d
    print >> f, cover_value
    f.flush()

    for name, values in d.items():
        for v in values:
            yield [name, v], cover_value
